File: too-many-cooks/lib/src/tools/status_tool.dart
/// Status tool - system overview.
library;

import 'package:dart_logging/dart_logging.dart';
import 'package:dart_node_mcp/dart_node_mcp.dart';
import 'package:nadz/nadz.dart';
import 'package:too_many_cooks/src/data/data.dart';
import 'package:too_many_cooks/src/types.dart';

/// Input schema for status tool (no inputs required).
const statusInputSchema = <String, Object?>{
  'type': 'object',
  'properties': <String, Object?>{},
};

/// Tool config for status.
const statusToolConfig = (
  title: 'Status',
  description: 'Get system overview: agents, locks, plans, messages',
  inputSchema: statusInputSchema,
  outputSchema: null,
  annotations: null,
);

/// Create status tool handler.
ToolCallback createStatusHandler(TooManyCooksDb db, Logger logger) =>
    (args, meta) async {
      final log = logger.child({'tool': 'status'});

      // Get agents
      final agentsResult = db.listAgents();
      if (agentsResult case Error(:final error)) {
        return _errorResult(error);
      }
      final String agents;
      switch (agentsResult) {
        case Success(:final value):
          agents = value.map(agentIdentityToJson).join(',');
        case Error(:final error):
          return _errorResult(error);
      }

      // Get locks
      final locksResult = db.listLocks();
      final String locks;
      switch (locksResult) {
        case Success(:final value):
          locks = value.map(fileLockToJson).join(',');
        case Error(:final error):
          return _errorResult(error);
      }

      // Get plans
      final plansResult = db.listPlans();
      final String plans;
      switch (plansResult) {
        case Success(:final value):
          plans = value.map(agentPlanToJson).join(',');
        case Error(:final error):
          return _errorResult(error);
      }

      // Get messages
      final messagesResult = db.listAllMessages();
      final String messages;
      switch (messagesResult) {
        case Success(:final value):
          messages = value.map(messageToJson).join(',');
        case Error(:final error):
          return _errorResult(error);
      }

      log.debug('Status queried');

      return (
        content: <Object>[
          textContent(
            '{"agents":[$agents],"locks":[$locks],'
            '"plans":[$plans],"messages":[$messages]}',
          ),
        ],
        isError: false,
      );
    };

CallToolResult _errorResult(DbError e) =>
    (content: <Object>[textContent(dbErrorToJson(e))], isError: true);

File: too-many-cooks/lib/src/tools/status_tool.ts
/// Status tool - system overview.

import type { Logger } from "../logger.js";
import {
  type TooManyCooksDb,
  type DbError,
  agentIdentityToJson,
  agentPlanToJson,
  dbErrorToJson,
  fileLockToJson,
  messageToJson,
} from "../data/data.js";
import {
  textContent,
  type CallToolResult,
  type ToolCallback,
} from "../types.js";

/** Input schema for status tool (no inputs required). */
export const STATUS_INPUT_SCHEMA = {
  type: "object",
  properties: {},
} as const;

/** Tool config for status. */
export const STATUS_TOOL_CONFIG = {
  title: "Status",
  description: "Get system overview: agents, locks, plans, messages",
  inputSchema: STATUS_INPUT_SCHEMA,
  outputSchema: null,
  annotations: null,
} as const;

/** Create status tool handler. */
export const createStatusHandler = (
  db: TooManyCooksDb,
  logger: Logger,
): ToolCallback =>
  async (_args, _meta) => {
    const log = logger.child({ tool: "status" });

    const agentsResult = db.listAgents();
    if (!agentsResult.ok) return makeErrorResult(agentsResult.error);
    const agents = agentsResult.value.map(agentIdentityToJson);

    const locksResult = db.listLocks();
    if (!locksResult.ok) return makeErrorResult(locksResult.error);
    const locks = locksResult.value.map(fileLockToJson);

    const plansResult = db.listPlans();
    if (!plansResult.ok) return makeErrorResult(plansResult.error);
    const plans = plansResult.value.map(agentPlanToJson);

    const messagesResult = db.listAllMessages();
    if (!messagesResult.ok) return makeErrorResult(messagesResult.error);
    const messages = messagesResult.value.map(messageToJson);

    log.debug("Status queried");

    return {
      content: [
        textContent(JSON.stringify({ agents, locks, plans, messages })),
      ],
      isError: false,
    };
  };

const makeErrorResult = (e: DbError): CallToolResult => ({
  content: [textContent(JSON.stringify(dbErrorToJson(e)))],
  isError: true,
});
